src/mango/src/mango_selector.erl

```diff
@@ -1054,4 +1061,29 @@ fields_nor_test() ->
     },
     ?assertEqual([<<"field1">>, <<"field2">>], fields_of(Selector2)).
 
+match_beginswith_test() ->
+    Doc =
+        {[
+            {<<"_id">>, <<"foo">>},
+            {<<"_rev">>, <<"bar">>},
+            {<<"user_id">>, 11}
+        ]},
+    Check = fun(Field, Prefix) ->
```

Comment:
Instead of copying, maybe it would be worth to lift the `Check` function to the module level and hook it up from here and `match_demo_test_/0` above (i.e. sharing).  This test case could be moved closed to that too.

Reply:
the `Check` function is test-specific so I'm not convinced it makes sense to share it. I don't think `match_demo_test_` is actually running any tests (this confused me for a while but note it calls `?_assertEqual` instead of `?assertEqual` which appears to be a no-op.

Reply:
The eunit test suite vs test vs generated test (a fun) is quite confusing. It's usually simpler to use basic test functions. For a suite, the setup/teardown with `?TDEF/?TDEF_FE` can work well.

Since we have two tests, it could be simpler to have a helper check function like @pgj suggested. The Doc it could be defined in a macro like ?TEST_DOC or re-defined inside the check match helper function, whichever looks better. Then, tests can just be a list of simple asserts:

```erlang
match_demo_test() ->
   ?assert(check_match({[{<<"user_id">>, 11}]} , ?TEST_DOC))),
   ?assertNot(check_match({[{<<"user_id">>, 1234}]}, ?TEST_DOC))),
   ...
```

So it looks like a simple function. If it fails on an assert we should get the exact line number and the expression where it fails.

Reply:
@willholley ah, I see -- I got probably mislead by the comment that was copied over from the other similarly-looking definition.  On the other hand, I tend to agree with @nickva regarding the actual definition itself.  Embedding the use of `Doc` variable internally makes the test cases a bit less explicit and hard to follow.  I am aware that this is coming from the other test case but perhaps we could cha(lle)nge the original structure here.

With regard to the no-op test case, maybe this is the time to either fix or scratch it?

Reply:
Ah I see now that it's not a no-op but a generated test. I'll try a version using the basic test functions as suggested.